Cache the value of torch_key in subproc

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151057

No need to recalculate torch_key in subprocs, lets pass it from main process.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

oulgen committed:
Cache the value of torch_key in subproc

[ghstack-poisoned]

oulgen commented on torch/_inductor/codecache.py:
```diff
@@ -662,12 +662,32 @@ def build_code_hash(
+        return _cache[0]
+
+    def set_val(val: bytes) -> None:
+        _cache.append(val)
```
err, this should _cache = [val], will fix

masnesral replied:
Maybe assert that it's empty too? (it should be empty, right?)

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -662,12 +662,32 @@ def build_code_hash(
 
 
-@functools.lru_cache(None)
+def torch_key_cache(func: Callable[[], bytes]) -> Callable[[], bytes]:
```
Nit: What's going on here exactly? Can you add a comment? I think this is just a straight-up reimplementation of functools.lru_cache, except with a set_val method?

oulgen replied:
yes it is. there's no way to inject a set_val method to lru_cache

masnesral reviewed (changes requested):
Mostly a question: not sure about the character encoding (but I admit I don't know a lot about such things.(

masnesral commented on torch/_inductor/compile_worker/subproc_pool.py:
```diff
@@ -115,6 +116,8 @@ def __init__(
         read_fd, subproc_write_fd = os.pipe()
         self.write_pipe = os.fdopen(write_fd, "wb")
         self.read_pipe = os.fdopen(read_fd, "rb")
+        # Cant use utf8 due to invalid start byte
```
Did my comment get dropped? Uhh ... I tried to say: I don't know hardly anything about character encodings, but is latin-1 always safe? Can it always encode this problematic start byte? And is it always safe to specify on the cmd line on all platforms. I wonder if it would be safer to base64 encode -- I think we do that in the cache key calculation, for example

oulgen committed:
None on "Cache the value of torch_key in subproc"


No need to recalculate torch_key in subprocs, lets pass it from main process.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -r

malfet commented:
@pytorchbot rebase -b main

oulgen commented:
@pytorchbot merge -f "github scheduler seems broken??"

oulgen committed:
Update

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge -f "github scheduler seems broken??"
